aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -143,6 +155,21 @@ Tensor where(const Tensor& condition, const Tensor& self, const Tensor& other) {
   return at::_s_where(b_condition, b_self, b_other);
 }
 
+Tensor where(const Tensor& condition, Scalar self, const Tensor& other) {
+  return at::where(condition, wrapped_scalar_tensor(self, other.device()), other);
+}
+
+Tensor where(const Tensor& condition, const Tensor& self, Scalar other) {
+  return at::where(condition, self, wrapped_scalar_tensor(other, self.device()));
+}
+
+Tensor where(const Tensor& condition, Scalar self, Scalar other) {
+  const auto device = condition.device();
+  const Tensor& other_t = wrapped_scalar_tensor(other, device);
```

Comment:
With apologies, I had the opportunity to think about this more and I think we'll need a tweak here.

When where is called with a tensor x scalar argument then type promotion will help us, but when it's just called with two scalar arguments it won't.

So we need to set the dtype of these tensors to be the appropriate default dtype (long, default_scalar_type, complex_default_scalar_type). This will make where(cond, .5, 1.5) produce a float tensor, for example, which is what we want. 

The easiest way to do this is probably a new function, "wrapped_default_dtype_scalar_tensor" or something named similarly, that inspects the scalars or the wrapped tensors and then assigns them the appropriate defaults. 

Does that make sense? 

Otherwise where(cond, .5, 1.5) will produce double tensors, instead of tensors in the default scalar type, even after implementing type promotion.

Reply:
Makes sense. 
I just have a question.
https://github.com/pytorch/pytorch/blob/d80e0c62bea9cc4f0b3c46edbcf892baee38b4c8/c10/core/DefaultDtype.cpp#L4-L22

Based on this file, I see that currently the default types are only verified available for `floating` and `complex` dtype.
Wanted to know if we have anything similar for `integral` types or should we just default integral to `Long`?

Reply:
Excellent question. You're correct, we have an implicit default integer default of long, and it can never be changed. So, as you predicted, you just want to make integral types to long.

Reply:
Thanks! Will do that.

Reply:
Is `torch.get_default_complex_dtype` exposed to Python. I haven't been able to find it.

Reply:
I don't think so (cc @anjali411). You could write a test helper that returns cfloat if the default scalar type is float32 and cdouble if the default scalar type is float64, though?

Reply:
Yup have done that in the updated test
```
complex_dtype = torch.complex64 if torch.float == torch.get_default_dtype() else torch.complex128
```

Reply:
no `torch.get_default_complex_dtype` not exposed to python yet. although I think we would like to have some way in python to get the corresponding complex dtype for real dtypes in future (since this logic is very commonly used)

Reply:
You could add a function to torch/testing/__init__.py for this.

Reply:
```python
def get_default_complex_dtype():
    return torch.complex64 if torch.float == torch.get_default_dtype() else torch.complex128
```
Should I add the above snippet in `init.py`?

I don't know of any other test which try to get the default complex dtype.